services/user-svc/internal/user/user.go

```diff
@@ -31,86 +34,93 @@ func NewServiceServer() *ServiceServer {
 
 func (s ServiceServer) CreateUser(ctx context.Context, req *pb.CreateUserRequest) (*pb.CreateUserResponse, error) {
 	log := zlog.Ctx(ctx)
-	userRepository := repositories.UserRepo(ctx)
+	userRepo := user_repo.New(hwdb.GetDB())
 
 	userID, err := uuid.Parse(req.Id)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	var user *models.User
-	if existingUser, _ := userRepository.GetUserById(userID); existingUser == nil {
-		avatarUrl := userRepository.GenerateDefaultAvatarUrl(userID.String())
-
-		user, err = userRepository.CreateUser(models.User{
-			ID: userID,
-			BaseUser: models.BaseUser{
-				Avatar:   avatarUrl,
-				Email:    req.Email,
-				Nickname: req.Nickname,
-				Name:     req.Name,
-			},
+	var createdUser user_repo.User
+	result, err := hwdb.Optional(userRepo.GetUserById)(ctx, userID)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	} else if result != nil {
+		return nil, status.Error(codes.InvalidArgument, "user already exists")
+	} else {
+		hash := sha256.Sum256([]byte(userID.String()))
+		avatarUrl := fmt.Sprintf("%s%s", "https://source.boringavatars.com/marble/128/", hex.EncodeToString(hash[:]))
+
+		createdUser, err = userRepo.CreateUser(ctx, user_repo.CreateUserParams{
+			ID:        userID,
+			Email:     req.Email,
+			Nickname:  req.Nickname,
+			Name:      req.Name,
+			AvatarUrl: &avatarUrl,
 		})
-
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 
 		userCreatedEvent := &events.UserCreatedEvent{
-			Id:       user.ID.String(),
-			Email:    user.Email,
-			Nickname: user.Nickname,
-			Name:     user.Name,
+			Id:       createdUser.ID.String(),
+			Email:    createdUser.Email,
+			Nickname: createdUser.Nickname,
+			Name:     createdUser.Name,
 		}
-
 		daprClient := common.MustNewDaprGRPCClient()
-
 		if err := common.PublishMessage(ctx, daprClient, "pubsub", "USER_CREATED", userCreatedEvent); err != nil {
 			log.Error().Err(err).Msg("could not publish message")
 		}
 	}
-
-	return &pb.CreateUserResponse{Id: user.ID.String()}, nil
+	return &pb.CreateUserResponse{Id: createdUser.ID.String()}, nil
 }
 
 func (s ServiceServer) ReadPublicProfile(ctx context.Context, req *pb.ReadPublicProfileRequest) (*pb.ReadPublicProfileResponse, error) {
-	userRepository := repositories.UserRepo(ctx)
+	userRepo := user_repo.New(hwdb.GetDB())
 
 	userID, err := uuid.Parse(req.Id)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	user, err := userRepository.GetUserById(userID)
+	user, err := hwdb.Optional(userRepo.GetUserById)(ctx, userID)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
+	} else if user == nil {
+		return nil, status.Error(codes.Internal, "record not found")
 	}
 
 	return &pb.ReadPublicProfileResponse{
 		Id:        user.ID.String(),
 		Name:      user.Name,
 		Nickname:  user.Nickname,
-		AvatarUrl: user.Avatar,
+		AvatarUrl: *user.AvatarUrl,
```

Comment:
AvatarUrl is not `NOT NULL`.
THIS WILL CAUSE NIL-DEREFERENCES!

Reply:
Hm true, thought because of the default value this shouldn't be a problem.
Forgot that it can theoretically be NULL if this was inserted intentionally, will fix that.